### PR TITLE
Add smoking information to maps

### PR DESCRIPTION
**Pull request by wetneb**
This is a first step towards displaying the smoking status of venues (#7577).

I added it as a type to make it searchable, following the treatment of `cuisine` and `wheelchair` information, but happy to keep it as a metadata attribute only if you prefer.

I couldn't figure out if I should regenerate `data/classificator.txt` and others myself in this PR, or if it should rather be left as a follow-up. Because I could not figure out how to regenerate those, I have made the changes manually there, by imitation. Happy to discard those changes if needed.

### CODE REVIEW

**Comment by vng** [comment]
- Keep only in metadata. Even if we want to have _smoking_ search, we should invent some other logic (and for cuisine, recycling, ...)
- Take only top-5 tags? https://taginfo.openstreetmap.org/keys/smoking#values
- Make only 3 enum values to show in PP: like Yes, No, Outside (with separated and isolated)?

**Comment by wetneb** [comment]
Thanks for the review!

> Make only 3 enum values to show in PP: like Yes, No, Outside (with separated and isolated)?

If we are to restrict to three values, then to me "separated" and "isolated" should rather be conflated with "yes", as they all indicate that *indoor* smoking is allowed.

**Comment by vng** [comment]
>If we are to restrict to three values, then to me "separated" and "isolated" should rather be conflated with "yes", as they all indicate that indoor smoking is allowed.

Ok, of course, we can keep all values, but better to decide when we will show some status in PP. Will it be icons or text?
@dvdmrtnz

File: data/mapcss-mapping.csv [hunk 1]
```diff
@@ -1683,3 +1683,6 @@ railway|station|subway|singapore;[railway=station][transport=subway][city=singap
 railway|subway_entrance|singapore;[railway=subway_entrance][city=singapore];;name;int_name;1595;
 railway|station|subway|boston;[railway=station][transport=subway][city=boston],[railway=station][station=subway][city=boston];;name;int_name;1596;
 railway|subway_entrance|boston;[railway=subway_entrance][city=boston];;name;int_name;1597;
+smoking|no;1598;
+smoking|outside;1599;
+smoking|yes;1600;
```
Comment (vng) [inline data/mapcss-mapping.csv:1688]: These changes are non needed. And others in ./data/

File: generator/osm2meta.cpp [hunk 2]
```diff
@@ -337,6 +337,17 @@ std::string MetadataTagProcessorImpl::ValidateAndFormat_local_ref(std::string co
...
 }
 
+std::string MetadataTagProcessorImpl::ValidateAndFormat_smoking(std::string v)
+{
+  strings::AsciiToLower(v);
+  if (v == "outside" || v == "yes" || v == "no")
+    return v;
+  // Process additional top tags.
+  if (v == "separated" || v == "isolated" || v == "dedicated")
+    return "yes";
```
Comment (vng) [inline generator/osm2meta.cpp:347]: I mean keep outside, yes, no, separated, isolated, dedicated values as-is.

But keep only 3 options when showing the status in the Place Page (not in this PR).

**Review by vng — CHANGES_REQUESTED** [review]

**Comment by dvdmrtnz** [comment]
> > If we are to restrict to three values, then to me "separated" and "isolated" should rather be conflated with "yes", as they all indicate that indoor smoking is allowed.
> 
> Ok, of course, we can keep all values, but better to decide when we will show some status in PP. Will it be icons or text? @dvdmrtnz

There’s these emojis that could be used: 🚬  for smoking=yes and 🚭 for smoking=no. However if we want to show more values it will have to be in text.

**Comment by wetneb** [comment]
I agree that an icon-based UI (like for wheelchair accessibility) is probably only doable to show a boolean. I made a mock-up of a text-based UI in the issue (#7577). I wonder to what extent we need to decide on one or another within this PR? Does it have any influence about how we store the metadata in the map?

If we are to use a boolean, then my intuition would be for it to answer the question of whether one can smoke indoors, as I would say that's the most relevant information, both for smokers and non-smokers. That would mean that`no` and `outside` would be displayed as non-smoking and `yes`, `separated`, `dedicated` would be displayed as smoking (`isolated` is a bit tricky, because although it allows indoor smoking, non-smokers will also be able to stay in a smoke-free space, so it feels a bit misleading to advertise the place as just "smoking").

If we go for this boolean scheme, then it would perhaps make sense to hide the icons on places that are outdoors-only. For instance in Germany we often have some "Biergarten" mapped separately from the bar they are associated with, and they are generally mapped with `smoking=yes`. Users could easily be confused by thinking they have looked up the entire place (including the indoor space) and infer from that that indoor smoking is allowed in the bar. If we want to implement this tweak then that's perhaps something that requires adaptations to the map generation (unless there is already sufficient metadata stored for the front-end to determine if a place is outdoors only).

**Comment by pastk** [comment]
For wheelchair we also use icons for boolean yes/no only, but when its "partial/limited" then a longer text is displayed - I find this logic fine and it could be applied for the smoking case as well.

**Comment by wetneb** [comment]
Let me know if there is anything that's needed on my side to get this merged. As I understand it, I need to wait for this to get merged before starting work on displaying this information in the Android app.

**Comment by vng** [comment]
To all interested, please write your opinion here. Do we need this tag in the data?
https://taginfo.openstreetmap.org/keys/smoking#overview

I have doubts because the world almost has become non-smoke. As you can see, no + outside are > 80%, so is it really necessary? And to draw it in PP?

**Comment by pastk** [comment]
> 
> I have doubts because the world almost has become non-smoke. As you can see, no + outside are > 80%, so is it really necessary? And to draw it in PP?

Its really hard to say without actually looking at the data in more detail...

Options:
- merge, generate new maps w/ that info, add UI (prototype at least), then beta test and rely on users' feedback to see if its useful in different regions
- make some OSM queries (overpass?) in different regions and analyze the tagging and correspondence to local laws and culture

Both seems like quite an effort..

**Comment by patepelo** [comment]
> To all interested, please write your opinion here. Do we need this tag in the data? https://taginfo.openstreetmap.org/keys/smoking#overview
> 
> I have doubts because the world almost has become non-smoke. As you can see, no + outside are > 80%, so is it really necessary? And to draw it in PP?

I'm not a smoker so I may be biased. I find the enrichment of POI info with tags necessary and OSM data in some regions is starting to be rich enough for many features. The problem is that we're approaching each feature individually. Without looking at longer horizon on how the place page would look like when we add other features such as:

Example:
- Dogs/Pets allowed
- Seating available outside
- Takeaway only
- Reservations accepted
- Cash payment only
- Draft beers available
- [...]
And the list could be very long. 

So yeah, we should start thinking of scaling this UI to start implementing these tags. 

Attached are some screenshots of examples from GMaps and Mapycz. 

- So I would oppose an icon approach for this feature, specially since over 80% of the values are "no" and start thinking of listed values.

- As per the options, I would also oppose "Smoking: Yes" and instead go for "- Smoking allowed"; "-Smoking forbidden", "Smoking allowed outside" (or "Smoking allowed partially"). As per adding the editor feature, I would leave it for a later issue to test first. Maybe adding "Smoke free space" or something alike would be more positive. 

- To the feature of being able to search for this feature, I can't give an opinion on how a user would search for that. Search for smoking allowed and the suddenly a mix of bars and restaurants? Would a user prefer a different place only because they're allowed to smoke? Would a different user prefer a smoke free place because the user is quitting smoking? Maybe these are rare cases. I wouldn't add them at the moment IMO.


I think it's very positive that these meta data is added and really appreciate the volunteering on solving it. This one is a bit complex due to the kind of tag it handles and how OM can resolve adding such info without unintentionally highlighting too much a feature that could potentially send a subliminal message for an unhealthy habit. 

Let's hear more from different perspectives. 😄 

P.S. Screenshots of feature lists:

Screenshot_20240320-215251
Screenshot_20240320-215258
Screenshot_20240320-215523

**Comment by biodranik** [comment]
It would be great to avoid promoting smoking in any way. We're working on _Organic_ Maps, right?

It would be great to collect important real-world use-cases first, before introducing this feature, so more users can get benefits from it.

**Comment by wetneb** [comment]
To react to @vng's concisely:

> I have doubts because the world almost has become non-smoke.

I wish you were right! But if you look at this [list of smoking bans from Wikipedia](https://en.wikipedia.org/wiki/List_of_smoking_bans) (in particular the map there) it's far from universal still.

>  As you can see, no + outside are > 80%, so is it really necessary? 

I'd be cautious about drawing conclusions from statistics on tag values, because it is probably far from a representative sample.

> It would be great to collect important real-world use-cases first, before introducing this feature, so more users can get benefits from it.

Let me give you my use case.

In Saxony (Germany) it is quite common for bars and other hospitality spaces to allow smoking. As a non-smoker I am really tired of ending up in bars where I am in contact with indoor smoking (be it `smoking=yes` or `smoking=separated`) because it makes me cough and my clothes stink afterwards. Finding out in advance if a place allows smoking is remarkably hard: it is rarely indicated on the website of the place (if it has one) and street imagery is rarely useful (signage is often absent or not very prominent). Google Maps does not give this information either, as far as I can tell.

So I believe offering this information would give OM a clear added value over Google Maps for people like me (and I think I am far from the only one here).

**Comment by pastk** [comment]
> 
> In Saxony (Germany) it is quite common for bars and other hospitality spaces to allow smoking. As a non-smoker I am really tired of ending up in bars where I am in contact with indoor smoking (be it `smoking=yes` or `smoking=separated`) because it makes me cough and my clothes stink afterwards. Finding out in advance if a place allows smoking is remarkably hard: it is rarely indicated on the website of the place (if it has one) and street imagery is rarely useful (signage is often absent or not very prominent). Google Maps does not give this information either, as far as I can tell.
> 

That is a great example and a use case, thanks!

I think we'd want to avoid situations like this:
    - most of the places has a smoking status tagged
    - most of the places around allow indoor smoking
    - then a user will end up seeing a "smoking allowed" text or emoji very often while its not being of much use (as its kind of status quo there)
    - what will be useful: seeing a "smoking free/forbidden" text/emoji on a few places (as they are not very common there)

or this:
    - most of the places has a smoking status tagged
    - most of the places around forbid indoor smoking
    - then a user will end up seeing a "smoking forbidden" text or emoji very often while its not being of much use
    - what will be useful: seeing a "smoking allowed" text/emoji on a few places (as they are not very common there)

If just a fraction of places has a smoking status tagged then it doesn't matter that much IMO.

Have you had a chance to research `smoking=*` stats in e.g. Saxony?

**Comment by wetneb** [comment]
> Have you had a chance to research smoking=* stats in e.g. Saxony?

In Saxony the rates of indoor smoking are pretty high, but I guess you probably want a broader overview to know if it's relevant to your user base at large. So I have started a more comprehensive review country by country:
https://wiki.openstreetmap.org/wiki/User:Pintoch/SmokingCountries#Table
Beyond the raw data I have also described the methodology a little - feedback welcome!

Out of this survey, I would say that smoking information is worth showing in those countries:
* Germany (or only in sub-regions of Germany if your country detector is sufficiently granular, sadly it's not the case for https://github.com/westnordost/countryboundaries which I used)
* Japan
* Indonesia
* Mexico
* Denmark
* Slovakia
* Croatia
* United Arab Emirates
* Serbia
* Belarus
* Bulgaria
* Vietnam
* Nepal
* Ecuador
* Kazakhstan
* Cuba
* Myanmar
* Morocco
* Tunisia
* Uzbekistan

But overall, I am not sure it's worth hard-coding a list of countries where this information should be shown or not. This is bound to be incomplete, prone to interpretation and disagreement and go out of date as the legislation evolve. I think it would be simpler to just show the information when it's provided in OSM - it's delegating the work of determining if it's locally relevant to OSM contributors themselves. After all, we have this:
https://wiki.openstreetmap.org/wiki/Good_practice#Don't_map_local_legislation_if_not_bound_to_specific_objects

If you think it's indeed worth including this information, I would be happy to update this PR and work on getting it merged.

**Comment by biodranik** [comment]
I would prefer to avoid indirectly advertizing smoking in Organic Maps.

**Comment by patepelo** [comment]
> I would prefer to avoid indirectly advertizing smoking in Organic Maps.

We could take the opportunity and flip the coin if we present it as a positive feature when a place is **smoke free** 🫁🍃

Think about hard smoking countries where legislation isn't helping and it'd be a positive thing to see in the PP.

I think the major stopping stone here is the previous discussion that we need to set a longer horizon for formatting the place page. We should make some prototypes of how we would fit everything together with all the tags available.

**Comment by pastk** [comment]
@wetneb Thanks for your research!

> 
> If you think it's indeed worth including this information, I would be happy to update this PR and work on getting it merged.

For sure we can release it as an alpha/experimental build and ask for users' feedback.
So there is always a probability the PR won't be accepted in the end.
If you're fine taking this risk then please go on with the implementation!

And regardless of your choice whether to continue or not - thanks a lot for your effort and raising this up!

**Comment by wetneb** [comment]
Given the mixed feedback I think this is unlikely to work out, so I will not work on this further.

**Pull request closed without merging**